File: problems/Python/76. Minimum Window Substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if t == "" or len(s) < len(t):
            return ""

        tmap = {}
        for c in t:
            tmap[c] = tmap.get(c, 0) + 1
        
        need = len(tmap)
        have = 0
        wmap = {}
        res = [-1, -1]
        rlen = float("inf")
        l = 0

        for r in range(len(s)):
            c = s[r]
            wmap[c] = wmap.get(c, 0) + 1

            if c in tmap and wmap[c] == tmap[c]:
                have += 1

            while have == need:
                if (r - l + 1) < rlen:
                    res = [l, r]
                    rlen = r - l + 1
                
                lc = s[l]
                wmap[lc] -= 1
                if lc in tmap and wmap[lc] < tmap[lc]:
                    have -= 1
                l += 1
        
        l, r = res
        return s[l:r+1] if rlen != float("inf") else ""

File: problems/C++/76. Minimum Window Substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        if (t.empty() || s.length() < t.length()) {
            return "";
        }
        
        unordered_map<char, int> tmap;
        for (char c : t) {
            tmap[c]++;
        }

        int l = 0, rlen = -1, ridx = -1;
        int need = tmap.size(), have = 0;
        unordered_map<char, int> wmap;

        for (int r = 0; r < s.length(); ++r) {
            char c = s[r];
            wmap[c]++;

            if (tmap.count(c) && wmap[c] == tmap[c]) {
                have++;
            }

            while (have == need) {
                if (rlen == -1 || r - l + 1 < rlen) {
                    rlen = r - l + 1;
                    ridx = l;
                }
                char lc = s[l];
                wmap[lc]--;
                if (tmap.count(lc) && wmap[lc] < tmap[lc]) {
                    have--;
                }
                l++;
            }
        }
        return (rlen == -1) ? "" : s.substr(ridx, rlen);
    }
};